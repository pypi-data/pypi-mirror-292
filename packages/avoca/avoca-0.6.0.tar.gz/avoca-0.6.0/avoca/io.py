from pathlib import Path

import pandas as pd


def to_csv(df: pd.DataFrame, path: Path, **kwargs) -> None:
    """Export a dataframe to a csv file."""
    df.to_csv(path, index=False, **kwargs)


def from_csv(path: Path) -> pd.DataFrame:
    """Read a csv file to a dataframe."""
    df = pd.read_csv(path, header=[0, 1])
    parse_dates = [("-", "datetime"), ("-", "start_datetime"), ("-", "end_datetime")]
    # Convert the datetime column to a datetime object
    for col in parse_dates:
        if col not in df.columns:
            continue
        df[col] = pd.to_datetime(df[col])
    return df
