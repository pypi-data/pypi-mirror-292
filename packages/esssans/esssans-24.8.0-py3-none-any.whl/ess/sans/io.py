# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)

import scipp as sc
import scippnexus as snx
from scippnexus.application_definitions import nxcansas

from .types import BackgroundSubtractedIofQ, OutFilename, RunNumber, RunTitle


def save_background_subtracted_iofq(
    *,
    iofq: BackgroundSubtractedIofQ,
    out_filename: OutFilename,
    run_number: RunNumber,
    run_title: RunTitle,
) -> None:
    """Save background-subtracted I(Q) histogram as an NXcanSAS file."""
    if iofq.bins is None:
        da = iofq.copy(deep=False)
    else:
        da = iofq.hist()
    if da.coords.is_edges('Q'):
        da.coords['Q'] = sc.midpoints(da.coords['Q'])
    with snx.File(out_filename, 'w') as f:
        f['sasentry'] = nxcansas.SASentry(title=run_title, run=run_number)
        f['sasentry']['sasdata'] = nxcansas.SASdata(da, Q_variances='resolutions')
