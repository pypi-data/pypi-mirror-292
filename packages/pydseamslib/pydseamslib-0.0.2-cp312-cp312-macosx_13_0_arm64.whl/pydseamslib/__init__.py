from pydseamslib import cyoda

__all__ = ["cyoda"]
