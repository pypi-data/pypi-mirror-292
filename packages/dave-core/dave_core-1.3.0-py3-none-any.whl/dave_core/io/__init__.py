# Copyright (c) 2022-2024 by Fraunhofer Institute for Energy Economics and Energy System Technology (IEE)
# Kassel and individual contributors (see AUTHORS file for details). All rights reserved.
# Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.

from dave_core.io.convert_format import *
from dave_core.io.file_io import *
from dave_core.io.io_utils import *
