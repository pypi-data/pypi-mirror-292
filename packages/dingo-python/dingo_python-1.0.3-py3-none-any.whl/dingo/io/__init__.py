from dingo.io.input import InputModel, RawInputModel
from dingo.io.output import SummaryModel, ErrorInfo
