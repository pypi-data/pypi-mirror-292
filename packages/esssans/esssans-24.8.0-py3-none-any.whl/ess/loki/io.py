# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
"""
Loading and merging of LoKI data.
"""

import scipp as sc
from ess.reduce import nexus

from ess.sans.types import (
    DetectorEventData,
    Filename,
    MonitorEventData,
    MonitorType,
    NeXusDetector,
    NeXusDetectorName,
    NeXusMonitor,
    NeXusMonitorName,
    RawSample,
    RawSource,
    RunType,
)


def load_nexus_sample(file_path: Filename[RunType]) -> RawSample[RunType]:
    return RawSample[RunType](nexus.load_sample(file_path))


def dummy_load_sample(file_path: Filename[RunType]) -> RawSample[RunType]:
    return RawSample[RunType](
        sc.DataGroup({'position': sc.vector(value=[0, 0, 0], unit='m')})
    )


def load_nexus_source(file_path: Filename[RunType]) -> RawSource[RunType]:
    return RawSource[RunType](nexus.load_source(file_path))


def load_nexus_detector(
    file_path: Filename[RunType], detector_name: NeXusDetectorName
) -> NeXusDetector[RunType]:
    # Events will be loaded later. Should we set something else as data instead, or
    # use different NeXus definitions to completely bypass the (empty) event load?
    dg = nexus.load_detector(
        file_path=file_path,
        detector_name=detector_name,
        selection={'event_time_zero': slice(0, 0)},
    )
    # The name is required later, e.g., for determining logical detector shape
    dg['detector_name'] = detector_name
    return NeXusDetector[RunType](dg)


def load_nexus_monitor(
    file_path: Filename[RunType], monitor_name: NeXusMonitorName[MonitorType]
) -> NeXusMonitor[RunType, MonitorType]:
    return NeXusMonitor[RunType, MonitorType](
        nexus.load_monitor(
            file_path=file_path,
            monitor_name=monitor_name,
            selection={'event_time_zero': slice(0, 0)},
        )
    )


def load_detector_event_data(
    file_path: Filename[RunType], detector_name: NeXusDetectorName
) -> DetectorEventData[RunType]:
    da = nexus.load_event_data(file_path=file_path, component_name=detector_name)
    return DetectorEventData[RunType](da)


def load_monitor_event_data(
    file_path: Filename[RunType], monitor_name: NeXusMonitorName[MonitorType]
) -> MonitorEventData[RunType, MonitorType]:
    da = nexus.load_event_data(file_path=file_path, component_name=monitor_name)
    return MonitorEventData[RunType, MonitorType](da)


providers = (
    load_nexus_detector,
    load_nexus_monitor,
    load_nexus_sample,
    load_nexus_source,
    load_detector_event_data,
    load_monitor_event_data,
)
"""Providers for loading single files."""
