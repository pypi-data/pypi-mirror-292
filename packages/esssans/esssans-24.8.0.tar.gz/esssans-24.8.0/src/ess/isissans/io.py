# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
"""
File loading functions for ISIS data, NOT using Mantid.
"""

from typing import NewType

import sciline
import scipp as sc

from ess.sans.types import (
    BackgroundRun,
    DirectBeam,
    DirectBeamFilename,
    Filename,
    MaskedDetectorIDs,
    PixelMaskFilename,
    RunType,
    SampleRun,
    TransmissionRun,
)

CalibrationFilename = NewType('CalibrationFilename', str)


def read_xml_detector_masking(filename: PixelMaskFilename) -> MaskedDetectorIDs:
    """Read a pixel mask from an XML file.

    The format is as follows, where the detids are inclusive ranges of detector IDs:

    .. code-block:: xml

        <?xml version="1.0"?>
        <detector-masking>
            <group>
                <detids>1400203-1400218,1401199,1402190-1402223</detids>
            </group>
        </detector-masking>

    Parameters
    ----------
    filename:
        Path to the XML file.
    """
    import xml.etree.ElementTree as ET  # nosec

    tree = ET.parse(filename)  # noqa: S314
    root = tree.getroot()

    masked_detids = []
    for group in root.findall('group'):
        for detids in group.findall('detids'):
            for detid in detids.text.split(','):
                detid = detid.strip()
                if '-' in detid:
                    start, stop = detid.split('-')
                    masked_detids += list(range(int(start), int(stop) + 1))
                else:
                    masked_detids.append(int(detid))

    return MaskedDetectorIDs(
        sc.array(dims=['detector_id'], values=masked_detids, unit=None, dtype='int32')
    )


class LoadedFileContents(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
    """Contents of a loaded file."""


def load_tutorial_run(filename: Filename[RunType]) -> LoadedFileContents[RunType]:
    return LoadedFileContents[RunType](sc.io.load_hdf5(filename))


def load_tutorial_direct_beam(filename: DirectBeamFilename) -> DirectBeam:
    return DirectBeam(sc.io.load_hdf5(filename))


def transmission_from_sample_run(
    data: LoadedFileContents[SampleRun],
) -> LoadedFileContents[TransmissionRun[SampleRun]]:
    """
    Use transmission from a sample run, instead of dedicated run.
    """
    return LoadedFileContents[TransmissionRun[SampleRun]](data)


def transmission_from_background_run(
    data: LoadedFileContents[BackgroundRun],
) -> LoadedFileContents[TransmissionRun[BackgroundRun]]:
    """
    Use transmission from a background run, instead of dedicated run.
    """
    return LoadedFileContents[TransmissionRun[BackgroundRun]](data)


providers = (read_xml_detector_masking,)
