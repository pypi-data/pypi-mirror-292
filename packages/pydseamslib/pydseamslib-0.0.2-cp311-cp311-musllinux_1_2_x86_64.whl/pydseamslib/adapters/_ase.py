from pydseamslib.cyoda import PointDouble, PointCloudDouble
import numpy as np
import ase


def map_LAMMPS_IDs_to_atomic_symbols(dict_map: dict(), atms_a: ase.Atoms):
    # Traversing by atom is faster than traversing atoms by key value in the map,
    # since there will be fewer map values than atoms
    # See https://github.com/d-SEAMS/PydSEAMSlib/pull/11#issuecomment-2243986351
    # for benchmarks
    for pt in atms_a:
        if pt.number in dict_map.keys():
            pt.symbol = dict_map[pt.number]
    return atms_a


def swap_key_value(my_dict_a):
    new_dict = {value: key for key, value in my_dict_a.items()}
    return new_dict


def list_pts(atms: ase.Atoms, lmp_a: dict, splicemask: list[bool], molids: list[int]):
    ptslist = []
    inslice_atms = atms[splicemask]
    map_lmp = swap_key_value(lmp_a)
    all_atm_id = np.asarray([x.index + 1 for x in atms])[splicemask]
    for idx, atm in enumerate(inslice_atms):
        pd = PointDouble()
        pd.x = atm.position[0]
        pd.y = atm.position[1]
        pd.z = atm.position[2]
        pd.c_type = map_lmp[atm.symbol]
        pd.inSlice = True  # Assumes that the atom provided are in the Slice.
        pd.atomID = all_atm_id[idx]
        pd.molID = molids[idx]
        ptslist.append(pd)
    return ptslist


def to_pointcloud(
    atms_a: ase.Atoms,
    lmp_a: dict,
    splicemask: list[bool],
    molids: list[int],
    fname,
    currentFrame=[1],
):
    _pcd = PointCloudDouble()
    inslice_atms = atms_a[splicemask]
    # TODO(ruhila): Assumes a rectangular box
    _pcd.box = np.diag(np.asarray(inslice_atms.get_cell()))
    _pcd.nop = len(inslice_atms)
    _pcd.pts = list_pts(atms_a, lmp_a, splicemask, molids)
    _pcd.idIndexMap = make_indexmap(atms_a, splicemask)
    _pcd.currentFrame = 1
    _pcd.boxLow = boxlow(fname)
    return _pcd


def make_indexmap(atms: ase.Atoms, splicemask: list[bool]):
    inslice_atms = atms[splicemask]
    atm_id = np.asarray([x.index + 1 for x in atms])[splicemask]
    ret = dict()
    for idx in range(len(inslice_atms)):
        key = int(atm_id[idx])
        ret[key] = idx
    return ret


def boxlow(fname):
    # Read the contents of the file into a variable
    opfile = open(fname)
    data = opfile.read()
    filelist = []
    for i in range(5, 8):
        parfile = data.splitlines()[i]
        needfile = parfile.split()[0]
        floatfile = float(needfile)
        filelist.append(floatfile)
    # Don't forget to close the file again
    opfile.close()
    return filelist
