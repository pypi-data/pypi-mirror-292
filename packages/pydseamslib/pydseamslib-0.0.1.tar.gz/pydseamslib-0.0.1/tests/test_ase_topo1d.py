import pprint
import shutil
import numpy as np

from approvaltests import verify, verify_file
from approvaltests.namer.default_namer_factory import NamerFactory
from ase.io import read as aseread


from pydseamslib import cyoda
from pydseamslib.adapters import _ase
from pathlib import Path

TRAJ = Path("subprojects/seams-core/input/traj/exampleTraj.lammpstrj")
strTRJ = str(TRAJ.absolute())


def test_nlist():
    # Get the frame
    resCloud = cyoda.readLammpsTrjreduced(
        filename=strTRJ,
        targetFrame=1,
        typeI=2,  # oxygenAtomType
        isSlice=False,
        coordLow=[0, 0, 0],
        coordHigh=[0, 0, 0],
    )

    # Construct a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )
    assert pcd.box == resCloud.box
    assert pcd.nop == resCloud.nop
    assert pcd.idIndexMap == resCloud.idIndexMap
    assert pcd.currentFrame == resCloud.currentFrame
    assert pcd.boxLow == resCloud.boxLow

    for idx in range(len(pcd.pts)):
        assert pcd.pts[idx].x == resCloud.pts[idx].x
        assert pcd.pts[idx].y == resCloud.pts[idx].y
        assert pcd.pts[idx].z == resCloud.pts[idx].z
        assert pcd.pts[idx].c_type == resCloud.pts[idx].c_type
        assert pcd.pts[idx].inSlice == resCloud.pts[idx].inSlice
        assert pcd.pts[idx].atomID == resCloud.pts[idx].atomID
        assert pcd.pts[idx].molID == resCloud.pts[idx].molID

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Validate the results
    verify(pprint.pformat(nl))


def test_hbnlist():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )

    # Validate the results
    verify(pprint.pformat(hl))


def test_hbnlist1():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )

    # Validate the results
    verify(pprint.pformat(hL))


def test_rings():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )
    # Gets every ring (non-primitives included)
    Rgs = cyoda.ringNetwork(
        nList=hL,
        maxDepth=6,
    )

    # Validate the results
    verify(pprint.pformat(Rgs))


def test_prisms():
    # Construct a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )
    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )
    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )
    # Gets every ring (non-primitives included)
    Rgs = cyoda.ringNetwork(
        nList=hL,
        maxDepth=6,
    )
    # Ensure the directory is not present before beginning
    outDir = "runOne/"  # / is important for the C++ engine..
    if Path(outDir).exists():
        shutil.rmtree(outDir)
    # Does the prism analysis for quasi-one-dimensional ice
    cyoda.prismAnalysis(
        path=outDir,
        rings=Rgs,
        nList=hL,
        yCloud=pcd,
        maxDepth=6,
        atomID=0,
        firstFrame=1,  # targetFrame
        currentFrame=1,  # frame
        doShapeMatching=False,
    )
    # Validate the run results
    verify_file(
        Path(f"{outDir}/topoINT/dataFiles/system-prisms-1.data"),
        options=NamerFactory.with_parameters("systemPrisms"),
    )
    verify_file(
        Path(f"{outDir}/topoINT/nPrisms.dat"),
        options=NamerFactory.with_parameters("nPrisms"),
    )
