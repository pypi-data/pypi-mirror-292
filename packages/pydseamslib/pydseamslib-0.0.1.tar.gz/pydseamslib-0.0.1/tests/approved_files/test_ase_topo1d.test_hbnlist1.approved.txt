[[0, 51, 81, 144],
 [1, 89, 223, 219],
 [2, 24, 17],
 [3, 107, 115, 97, 235],
 [4, 220, 241],
 [5, 144, 242, 62, 93],
 [6, 20, 30, 11, 74],
 [7, 88, 101, 167],
 [8, 63, 206, 68],
 [9, 234, 135],
 [10, 29, 138, 117, 146],
 [11, 6, 131, 19, 92],
 [12, 141, 161, 27, 55],
 [13, 110, 247, 86, 188],
 [14, 38, 149, 208],
 [15, 71, 74, 30, 178],
 [16, 21, 163, 99, 210],
 [17, 2, 176, 48, 111],
 [18, 184, 190, 40, 129],
 [19, 11, 156, 41, 85],
 [20, 6, 245, 249, 94],
 [21, 16, 128, 167, 239],
 [22, 184, 190, 63, 78],
 [23, 121, 209, 31, 170],
 [24, 2, 221, 236, 111],
 [25, 143, 171, 44, 95],
 [26, 95, 196, 44, 162],
 [27, 12, 164, 117, 146],
 [28, 31, 235, 209],
 [29, 10, 59, 238, 223],
 [30, 6, 15, 131, 94],
 [31, 28, 23, 81, 73],
 [32, 110, 247, 106, 244],
 [33, 188, 244, 110, 225],
 [34, 203, 160],
 [35, 82, 218, 37, 53],
 [36, 61, 90, 103, 191],
 [37, 35, 118, 80, 232],
 [38, 14, 185, 230, 154],
 [39, 182, 197, 87, 134],
 [40, 18, 133, 58, 181],
 [41, 19, 43, 131, 140],
 [42, 72, 70, 147],
 [43, 41, 85, 153, 108],
 [44, 25, 26, 70, 147],
 [45, 130, 181, 155, 237],
 [46, 142, 198, 229],
 [47, 70, 137, 124, 165],
 [48, 17, 176, 205, 226],
 [49, 52, 150, 91, 98],
 [50, 53, 218, 194, 204],
 [51, 0, 54, 60, 193],
 [52, 49, 87, 134, 89],
 [53, 50, 35, 232],
 [54, 51, 183, 81],
 [55, 12, 164, 207, 214],
 [56, 220, 62, 93],
 [57, 84, 228, 136, 163],
 [58, 40, 78, 190, 233],
 [59, 29, 146, 207, 164],
 [60, 51, 144, 183],
 [61, 36, 86, 246, 76],
 [62, 5, 56, 142, 175],
 [63, 8, 22, 155, 130],
 [64, 169, 201, 119],
 [65, 101, 167, 132, 168],
 [66, 145, 237, 149, 227],
 [67, 80, 186, 207, 214],
 [68, 8, 227, 155, 237],
 [69, 171, 95, 222],
 [70, 47, 42, 44, 162],
 [71, 15, 92, 140, 131],
 [72, 42, 162, 166, 196],
 [73, 31, 235, 107, 121],
 [74, 15, 6, 249, 92],
 [75, 109, 186, 80, 118],
 [76, 61, 90, 172, 180],
 [77, 135, 165, 137, 234],
 [78, 58, 22, 155, 181],
 [79, 106, 112, 110, 225],
 [80, 67, 37, 75, 139],
 [81, 31, 0, 54],
 [82, 35, 100, 118, 177],
 [83, 98, 157, 136, 228],
 [84, 57, 114, 211, 197],
 [85, 43, 19, 125, 169],
 [86, 61, 13, 103, 126],
 [87, 52, 39, 157, 98],
 [88, 7, 210, 212],
 [89, 1, 52, 150, 148],
 [90, 36, 76, 102, 233],
 [91, 49, 134, 182],
 [92, 71, 11, 74, 156],
 [93, 5, 56, 193],
 [94, 20, 30, 112, 178],
 [95, 26, 25, 69, 189],
 [96, 149, 120, 145],
 [97, 3, 203, 152, 158],
 [98, 83, 49, 87],
 [99, 16, 239, 114],
 [100, 82, 194, 159],
 [101, 7, 65, 212, 231],
 [102, 90, 180, 191, 133],
 [103, 86, 36, 127, 247],
 [104, 227, 120, 145],
 [105, 119, 122, 200, 201],
 [106, 79, 32, 213, 243],
 [107, 3, 73, 158, 160],
 [108, 43, 122, 125],
 [109, 75, 139, 159, 161],
 [110, 13, 32, 33, 79],
 [111, 17, 24, 135, 202],
 [112, 79, 94, 213, 245],
 [113, 188, 244, 127, 247],
 [114, 84, 99, 216, 163],
 [115, 3, 116, 160, 203],
 [116, 115, 143, 222, 171],
 [117, 10, 27, 219, 238],
 [118, 37, 82, 75, 159],
 [119, 105, 64, 204, 153],
 [120, 96, 104, 154, 185],
 [121, 23, 73, 152],
 [122, 105, 108, 153, 195],
 [123, 208, 226, 230],
 [124, 47, 162, 166],
 [125, 85, 108, 195, 156],
 [126, 86, 246, 127, 188],
 [127, 103, 113, 126, 191],
 [128, 21, 163, 216],
 [129, 18, 133, 172, 180],
 [130, 45, 63, 206, 184],
 [131, 11, 30, 41, 71],
 [132, 65, 231, 248],
 [133, 40, 129, 102, 233],
 [134, 52, 91, 39, 148],
 [135, 77, 9, 111, 176],
 [136, 57, 83, 211],
 [137, 47, 77, 166],
 [138, 10, 215, 219, 223],
 [139, 109, 80, 232],
 [140, 71, 41, 156],
 [141, 12, 146, 207, 186],
 [142, 46, 62, 241, 220],
 [143, 25, 116, 147, 189],
 [144, 5, 60, 0, 170],
 [145, 66, 96, 104, 206],
 [146, 59, 141, 10, 27],
 [147, 143, 42, 44, 196],
 [148, 134, 89, 219, 215],
 [149, 14, 96, 66, 154],
 [150, 49, 89, 215, 223],
 [151, 224, 199, 231],
 [152, 121, 97, 209, 235],
 [153, 43, 122, 119, 169],
 [154, 149, 38, 120, 227],
 [155, 63, 78, 45, 68],
 [156, 19, 140, 92, 125],
 [157, 83, 87, 197, 211],
 [158, 107, 97],
 [159, 109, 118, 100, 232],
 [160, 115, 34, 107, 171],
 [161, 12, 109, 214, 186],
 [162, 72, 124, 26, 70],
 [163, 16, 128, 57, 114],
 [164, 27, 55, 59, 238],
 [165, 77, 47, 176],
 [166, 72, 137, 124, 234],
 [167, 21, 65, 7, 210],
 [168, 65, 248, 210, 212],
 [169, 64, 153, 85, 195],
 [170, 144, 23, 183],
 [171, 25, 69, 116, 160],
 [172, 76, 129, 190, 233],
 [173, 177, 200, 179, 201],
 [174, 187, 199, 248],
 [175, 62, 192, 242],
 [176, 17, 48, 135, 165],
 [177, 173, 82, 194, 218],
 [178, 15, 94, 213, 249],
 [179, 173, 204, 218],
 [180, 102, 76, 129],
 [181, 45, 40, 78, 184],
 [182, 39, 91],
 [183, 54, 60, 170, 209],
 [184, 18, 22, 130, 181],
 [185, 38, 120, 236],
 [186, 67, 75, 141, 161],
 [187, 174, 229, 198],
 [188, 33, 113, 13, 126],
 [189, 143, 95, 196, 222],
 [190, 18, 22, 58, 172],
 [191, 102, 36, 127, 246],
 [192, 175, 193],
 [193, 192, 51, 93, 242],
 [194, 100, 177, 50],
 [195, 125, 169, 122, 201],
 [196, 26, 189, 72, 147],
 [197, 39, 157, 84, 228],
 [198, 46, 187, 199, 220],
 [199, 174, 198, 151, 217],
 [200, 173, 105, 204],
 [201, 64, 195, 105, 173],
 [202, 111, 226, 221],
 [203, 34, 97, 115],
 [204, 119, 200, 50, 179],
 [205, 48, 234],
 [206, 8, 130, 145, 237],
 [207, 59, 141, 55, 67],
 [208, 123, 14, 221, 236],
 [209, 23, 152, 28, 183],
 [210, 88, 167, 16, 168],
 [211, 84, 157, 136, 216],
 [212, 101, 88, 168, 217],
 [213, 106, 112, 178, 240],
 [214, 161, 55, 67],
 [215, 138, 150, 148],
 [216, 114, 211, 128, 239],
 [217, 199, 212, 248, 231],
 [218, 35, 50, 177, 179],
 [219, 138, 148, 1, 117],
 [220, 4, 56, 142, 198],
 [221, 24, 208, 202],
 [222, 116, 69, 189],
 [223, 1, 150, 29, 138],
 [224, 151, 229],
 [225, 33, 79, 243, 245],
 [226, 123, 202, 48],
 [227, 68, 104, 66, 154],
 [228, 57, 197, 83],
 [229, 187, 46, 224, 241],
 [230, 38, 123],
 [231, 101, 132, 151, 217],
 [232, 53, 159, 37, 139],
 [233, 90, 133, 58, 172],
 [234, 9, 166, 77, 205],
 [235, 28, 73, 3, 152],
 [236, 24, 185, 208],
 [237, 66, 206, 45, 68],
 [238, 29, 164, 117],
 [239, 99, 216, 21],
 [240, 213, 243, 245, 249],
 [241, 142, 4, 229],
 [242, 5, 175, 193],
 [243, 240, 106, 225],
 [244, 33, 113, 32],
 [245, 20, 240, 112, 225],
 [246, 61, 126, 191],
 [247, 13, 32, 103, 113],
 [248, 168, 217, 132, 174],
 [249, 20, 74, 178, 240]]
